sklearn/metrics/pairwise.py

```diff
@@ -992,24 +992,28 @@ def detect_categorical_features(X, categorical_features=None):
 
     Returns
     -------
-    categorical_features_mask, numerical_features_mask
-     : ndarray, shape (n_features)
+    categorical_features_mask : ndarray, shape (n_features)
+
     """
-    _, n_cols = np.shape(X)
     # Automatic detection of categorical features
     if categorical_features is None:
-        categorical_features = np.zeros(n_cols, dtype=bool)
-        for col in range(0, n_cols):
-            if not np.issubdtype(type(X[0][col]), np.number):
+        categorical_features = np.zeros(np.shape(X)[1], dtype=bool)
+
+        def detect_cat(x):
+            return x != np.nan and not np.issubdtype(type(x), np.number)
+
+        for col in range(np.shape(X)[1]):
+            f_test = np.frompyfunc(detect_cat, 1, 1)
```

Comment:
frompyfunc is not efficient, and not, I think, necessary. Rather, you want to check the type of the first index that is not a NaN.

Reply:
Sorry, there is a wrong commit here, I'll push the correction, but we can assume for this private function we'll be receiving only numpy arrays.

Nevertheless, I don't feel comfortable anymore to check only the first row (that was a great catch, thanks that review), for this reason i'm traversing the entire set of data for each column. Definitely not a good strategy, as we just need to check the first not nan element. I'll push a better proposal together with Columntransformer usage.